//
// KVBlockChain data service
//

syntax = "proto3";
package com.digitalasset.kvbc;
option java_package = "com.digitalasset.kvbc";

service DataService {
  // Read transaction to read a set of keys.
  // TODO(JM): streaming instead? see keyvaluestore example in
  // grpc.
  rpc ReadTransaction (ReadTransactionRequest) returns (ReadTransactionResponse) {}

  rpc GetLatestBlockId (GetLatestBlockIdRequest) returns (BlockId) {}
}

message BlockId {
  int64 block_id = 1;
}

message GetLatestBlockIdRequest {
}

message ReadTransactionRequest {
  repeated bytes keys = 1;

  // Optional block identifier. If left out the read is done on latest block.
  int64 block_id = 2;
}

message ReadTransactionResponse {
  // The block on which the reads were executed.
  // FIXME(JM): Each read may come from a different
  // block...
  int64 block_id = 1;

  repeated KeyValuePair results = 2;
}

message KeyValuePair {
  bytes key = 1;
  bytes value = 2;
}
